Compile 0.14.72 failed: tcptunserver.cpp:48:7: error: use of undeclared identifier 'getsockname'

I currently try to prepare version 0.14.72 for FreeBSD, but it fails to compile.
Previous used version 0.14.66 was working fine.

Is there any new dependency?

Error messages are:
```
./link.h:350:21: note: overridden virtual function is here
  350 |   virtual FilterPtr findFilter(std::string name, bool skip_me = false)
      |                     ^
tcptunserver.cpp:48:7: error: use of undeclared identifier 'getsockname'
   48 |   if (getsockname(fd, (struct sockaddr *)&localSocketAddress, &len) != 0 ||
      |       ^
tcptunserver.cpp:50:40: error: use of undeclared identifier 'AF_INET'
   50 |       localSocketAddress.sin_family != AF_INET)
      |                                        ^
tcptunserver.cpp:515:3: error: use of undeclared identifier 'setsockopt'
  515 |   setsockopt(cfd, IPPROTO_TCP, TCP_NODELAY, &val, sizeof (val));
      |   ^
tcptunserver.cpp:578:21: error: use of undeclared identifier 'AF_INET'
  578 |   addr.sin_family = AF_INET;
      |                     ^
tcptunserver.cpp:582:15: error: use of undeclared identifier 'AF_INET'
  582 |   fd = socket(AF_INET, SOCK_STREAM, 0);
      |               ^
tcptunserver.cpp:582:24: error: use of undeclared identifier 'SOCK_STREAM'
  582 |   fd = socket(AF_INET, SOCK_STREAM, 0);
      |                        ^
tcptunserver.cpp:589:18: error: use of undeclared identifier 'SOL_SOCKET'
  589 |   setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof (reuse));
      |                  ^
tcptunserver.cpp:589:30: error: use of undeclared identifier 'SO_REUSEADDR'
  589 |   setsockopt(fd, SOL_SOCKET, SO_REUSEADDR, &reuse, sizeof (reuse));
      |                              ^
tcptunserver.cpp:591:7: error: use of undeclared identifier 'bind'
  591 |   if (bind(fd, (struct sockaddr *) &addr, sizeof (addr)) == -1)
      |       ^
tcptunserver.cpp:597:7: error: use of undeclared identifier 'listen'
  597 |   if (listen(fd, 10) == -1)
      |       ^
tcptunserver.cpp:656:21: error: use of undeclared identifier 'AF_LOCAL'
  656 |   addr.sun_family = AF_LOCAL;
      |                     ^
tcptunserver.cpp:659:15: error: use of undeclared identifier 'AF_LOCAL'
  659 |   fd = socket(AF_LOCAL, SOCK_STREAM, 0);
      |               ^
tcptunserver.cpp:659:25: error: use of undeclared identifier 'SOCK_STREAM'
  659 |   fd = socket(AF_LOCAL, SOCK_STREAM, 0);
      |                         ^
tcptunserver.cpp:666:7: error: use of undeclared identifier 'bind'
  666 |   if (bind(fd, (struct sockaddr *) &addr, sizeof (addr)) == -1)
      |       ^
tcptunserver.cpp:673:15: error: use of undeclared identifier 'connect'
  673 |           if (connect(fd, (struct sockaddr *) &addr, sizeof (addr)) == 0)
      |               ^
tcptunserver.cpp:686:19: error: use of undeclared identifier 'bind'
  686 |               if (bind(fd, (struct sockaddr *) &addr, sizeof (addr)) == -1)
      |                   ^
tcptunserver.cpp:700:7: error: use of undeclared identifier 'listen'
  700 |   if (listen(fd, 10) == -1)
      |       ^
tcptunserver.cpp:735:7: error: use of undeclared identifier 'listen'
  735 |   if (listen(fd, 10) == -1)
      |       ^
22 warnings and 18 errors generated.
gmake[3]: *** [Makefile:573: tcptunserver.o] Error 1
gmake[3]: *** Waiting for unfinished jobs....
warning: unknown warning option '-Wno-subobject-linkage' [-Wunknown-warning-option]
```

A full build log can be found here:
https://pkg.fechner.net/data/142amd64-gitlab/2025-01-22_12h26m41s/logs/errors/knxd-0.14.72.log

Thanks a lot for any hint.

On 22.01.25 11:35, Matthias Fechner wrote:
> tcptunserver.cpp:48:7: error: use of undeclared identifier 'getsockname'

Well, tcptunserver.cpp is new and appears not to have a complete set of 
#include files (as far as *BSD is concerned).

Please grep through /usr/include, find the header that defines 
getsockname, add that to the list at the top, and send us a patch.

-- 
-- mit freundlichen Grüßen
-- 
-- Matthias Urlichs

--- a/src/libserver/tcptunserver.cpp
+++ b/src/libserver/tcptunserver.cpp
@@ -23,6 +23,7 @@
 
 #include <unistd.h>
 #include <netinet/tcp.h>
+#include <sys/socket.h>
 #include <sys/un.h>
 #include <arpa/inet.h>
 

